quinn-proto/src/frame.rs

```diff
@@ -790,14 +947,81 @@ impl From<UnexpectedEnd> for IterErr {
     }
 }
 
+pub struct AckTimestampIter<'a> {
+    timestamp_basis: u64,
+    timestamp_exponent: u64,
+    timestamp_instant_basis: Instant,
+    data: &'a [u8],
+
+    deltas_remaining: usize,
+    first: bool,
+    next_pn: u64,
+}
+
+impl<'a> AckTimestampIter<'a> {
+    fn new(largest: u64, basis_instant: Instant, exponent: u64, mut data: &'a [u8]) -> Self {
+        // We read and throw away the Timestamp Range Count value because
+        // it was already used to properly slice the data.
+        let _ = data.get_var().unwrap();
+        AckTimestampIter {
+            timestamp_basis: 0,
+            timestamp_exponent: exponent,
+            timestamp_instant_basis: basis_instant,
+            data,
+            deltas_remaining: 0,
+            first: true,
+            next_pn: largest,
+        }
+    }
+}
+
+impl<'a> Iterator for AckTimestampIter<'a> {
+    type Item = PacketTimestamp;
+    fn next(&mut self) -> Option<Self::Item> {
+        if !self.data.has_remaining() {
+            debug_assert!(
+                self.deltas_remaining == 0,
+                "timestamp delta remaining should be 0"
+            );
+            return None;
+        }
+        if self.deltas_remaining == 0 {
+            let gap = self.data.get_var().unwrap();
+            self.deltas_remaining = self.data.get_var().unwrap() as usize;
+            if self.first {
+                self.next_pn -= gap;
+            } else {
+                self.next_pn -= gap + 2;
+            }
+        } else {
+            self.next_pn -= 1;
+        }
+
+        let delta = self.data.get_var().unwrap();
```

Comment:
Why is this safe?

Reply:
> I think it's considered safe because this is defined in the spec.

What's in the spec doesn't matter. Quinn must not panic just because a peer is non-conformant. That would be a DoS vulnerability.

I imagine this is safe because `scan_ack_timestamps_blocks` validated the ranges already. If that was the intent, document it, even if we've failed to do so elsewhere.